henchman/plan.go

```diff
@@ -39,10 +42,170 @@ func localhost() *Machine {
+			"remotePath": remoteModDir,
+			"host":       machine.Hostname,
+		}, "While creating dir")
+	}
+
+	// throw a check the check sum stuff in here somewhere
+	// transfer tar module
+	if err := machine.Transport.Put(target, remoteModDir, "dir"); err != nil {
+		return HenchErr(err, log.Fields{
+			"remotePath": remoteModDir,
+			"host":       machine.Hostname,
+		}, "While transfering tar")
+	}
+
+	// untar the modules
+	cmd := fmt.Sprintf("tar -xvf %s -C %s", remoteModDir+target, remoteModDir)
+	if _, err := machine.Transport.Exec(cmd, nil, false); err != nil {
+		return HenchErr(err, log.Fields{
+			"remotePath": remoteModDir,
+			"host":       machine.Hostname,
+		}, "While untarring")
+	}
+
+	// remove tar file
+	cmd = fmt.Sprintf("/bin/rm %s", remoteModDir+target)
+	if _, err := machine.Transport.Exec(cmd, nil, false); err != nil {
+		return HenchErr(err, log.Fields{
+			"remotePath": remoteModDir,
+			"host":       machine.Hostname,
+		}, "While removing tar in remote path")
+	}
+
+	return nil
+}
+
+// Moves all modules to each host
+func (plan *Plan) Setup(machines []*Machine) error {
+	log.WithFields(log.Fields{
+		"plan":         plan.Name,
+		"num machines": len(machines),
+	}).Info("Setting up plan")
+
+	// initialize set to hold module names
+	modSet := make(map[string]bool)
+
+	// get the curdir and move to location of modules
+	curDir, err := os.Getwd()
+	if err != nil {
+		return HenchErr(err, log.Fields{
+			"plan": plan.Name,
+		}, "")
+	}
+
+	// create the tar file to be filled
+	// create the writer to tar file
+	target := "modules.tar"
+	tarfile, err := os.Create(target)
+	if err != nil {
+		return HenchErr(err, log.Fields{
+			"plan":   plan.Name,
+			"target": target,
+		}, "")
+	}
+	tarball := tar.NewWriter(tarfile)
+
+	// gather all modules needed and verify they exist
+	// NOTE: just transfer everything to local
+	for _, task := range plan.Tasks {
+		/*
+			if _, ok := localSet[task.Module.Name]; !ok && task.Local {
+				localSet[task.Module.Name] = struct{}{}
+			} else if _, ok := modSet[task.Module.Name]; !ok {
+		*/
+		if _, ok := modSet[task.Module.Name]; !ok {
+			if _, err := task.Module.Resolve(); err != nil {
+				return HenchErr(err, log.Fields{
+					"plan": plan.Name,
+					"task": task.Name,
+				}, "")
+			}
+			modSet[task.Module.Name] = false
+		}
+	}
+
+	// tars all modules needed on remote machines
+	// NOTE: maybe we gotta zip them too
+	for _, modPath := range ModuleSearchPath {
+
+		// change to mod path
+		os.Chdir(modPath)
+
+		// add all modules in every search path
+		for modName, added := range modSet {
+
+			// if module has not been tarred add it
+			if !added {
+				info, err := os.Stat(modName)
+				if err != nil {
+					Debug(log.Fields{
+						"process": "getting module info",
+						"modPath": modPath,
+						"module":  modName,
+						"error":   err.Error(),
+					}, "Module not found")
+				} else {
+					if info.IsDir() {
+						if err = tarDir(modName, tarball); err != nil {
```

Comment:
I was thinking about making the tarring of modules a go func as you suggested.  I wonder if it would be better to have a channel manage writing to the modules.tar or just use a mutex lock.